internal/crypto/cert.go

```diff
@@ -37,32 +48,104 @@ func CNFromDeviceFingerprint(fingerprint string) (string, error) {
 	return DeviceCommonNamePrefix + fingerprint, nil
 }
 
+func CertFile(name string) string {
+	return filepath.Join(config.CertificateDir(), name+".crt")
+}
+
+func KeyFile(name string) string {
+	return filepath.Join(config.CertificateDir(), name+".key")
+}
+
 type TLSCertificateConfig oscrypto.TLSCertificateConfig
 
-type CA struct {
-	Config *TLSCertificateConfig
+type CA interface {
+	GetId() string
+	GetConfig() *TLSCertificateConfig
+	EnsureServerCertificate(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, bool, error)
+	MakeAndWriteServerCert(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, error)
+	MakeServerCert(hostnames []string, expiryDays int, fns ...CertificateExtensionFunc) (*TLSCertificateConfig, error)
+	signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error)
+	EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error)
+	MakeClientCertificate(certFile, keyFile string, subject string, expiryDays int) (*TLSCertificateConfig, error)
+	IssueRequestedClientCertificate(csr *x509.CertificateRequest, expirySeconds int) ([]byte, error)
+}
+
+
+
+var caMutex = &sync.Mutex{}
+var caList map[string]CA = make(map[string]CA)
```

Comment:
How about we encapsulate this in a struct vs exposing via global? It is generally good practice to avoid globals if possible from a maintainability and testing perspective.

``` suggestion
type CARegistry struct {
   mu sync.Mutex
   certs map[string]CA
}

func (m *CARegistry) Get(key string) (CA, bool)
...
func (m *CARegistry) Put(key string, ca CA)
```

Reply:
Yeah fine. I know it is k8 style. It am not a fan of piling init args and carrying them across the whole codebase for something that is obviously single instance and global. But so be it.

Reply:
The choice is yours I am just pointing out an observation that allowing any function/method access via global could be problematic down the road.